traffic_ops/traffic_ops_golang/topology/topologies.go

```diff
@@ -0,0 +1,475 @@
+
+func (topology TOTopology) GetKeyFieldsInfo() []api.KeyFieldInfo {
+	return []api.KeyFieldInfo{{"name", api.GetStringKey}}
+}
+
+func (topology *TOTopology) GetType() string {
+	return "topology"
+}
+
+func (topology *TOTopology) Validate() error {
+	nameRule := validation.NewStringRule(tovalidate.IsAlphanumericUnderscoreDash, "must consist of only alphanumeric, dash, or underscore characters.")
+	rules := validation.Errors{}
+	rules["name"] = validation.Validate(topology.Name, validation.Required, nameRule)
+
+	nodeCount := len(*topology.Nodes)
+	if nodeCount < 1 {
+		rules["length"] = fmt.Errorf("must provide 1 or more node, %v found", nodeCount)
+	}
+	cacheGroups := make([]*tc.CacheGroupNullable, nodeCount)
+	var err error
+	cacheGroupsExist := true
+	for index := 0; index < nodeCount; index++ {
+		node := (*topology.Nodes)[index]
+		rules[fmt.Sprintf("node %v parents size", index)] = validation.Validate((*node).Parents, validation.Length(0, 2))
+		rules[fmt.Sprintf("node %v duplicate parents", index)] = checkForDuplicateParents(topology.Nodes, index)
+		rules[fmt.Sprintf("node %v self parent", index)] = checkForSelfParents(topology.Nodes, index)
+		if cacheGroups[index], err = cachegroup.GetCacheGroupByName((*node).Cachegroup, &topology.APIInfoImpl); err != nil {
+			rules[fmt.Sprintf("cachegroup %v not found", index)] = fmt.Errorf("error getting cachegroup %v: %v", (*node).Cachegroup, err.Error())
+			cacheGroupsExist = false
+		}
+	}
+	rules["duplicate cachegroup name"] = checkUniqueCacheGroupNames(topology.Nodes)
+
+	if cacheGroupsExist {
+		for index := 0; index < nodeCount; index++ {
+			rules[fmt.Sprintf("parent '%v' edge type", (*topology.Nodes)[index].Cachegroup)] = checkForEdgeParents(topology.Nodes, &cacheGroups, index)
+		}
+
+		for _, leafMid := range *checkForLeafMids(topology.Nodes, &cacheGroups) {
+			rules[fmt.Sprintf("node %v leaf mid", (*leafMid).Cachegroup)] = fmt.Errorf("cachegroup %v's type is %v; it cannot be a leaf (it must have at least 1 child)", (*leafMid).Cachegroup, tc.MidCacheGroupType)
+		}
+	}
+	rules["topology cycles"] = checkForCycles(topology.Nodes)
+
+	errs := tovalidate.ToErrors(rules)
+	return util.JoinErrs(errs)
+}
+
+// Implementation of the Identifier, Validator interface functions
+func (topology TOTopology) GetKeys() (map[string]interface{}, bool) {
+	return map[string]interface{}{"name": topology.Name}, true
+}
+
+func (topology *TOTopology) SetKeys(keys map[string]interface{}) {
+	topology.Name, _ = keys["name"].(string)
+}
+
+func (topology *TOTopology) GetAuditName() string {
+	return topology.Name
+}
+
+func (topology *TOTopology) Create() (error, error, int) {
+	tx := topology.APIInfo().Tx.Tx
+	err := tx.QueryRow(insertQuery(), topology.Name, topology.Description).Scan(&topology.Name, &topology.Description, &topology.LastUpdated)
+	if err != nil {
+		userErr, sysErr, errCode := api.ParseDBError(err)
+		return userErr, sysErr, errCode
```

Comment:
seems like you could just `return api.ParseDBError(err)`

Reply:
Returning directly in f39f3349a.